File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/listeners/bungee/JoinLeaveListener.java
package nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee;

import com.google.common.base.Preconditions;
import lombok.RequiredArgsConstructor;
import net.dv8tion.jda.core.entities.Member;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.event.PostLoginEvent;
import net.md_5.bungee.api.plugin.Listener;
import net.md_5.bungee.event.EventHandler;

import net.md_5.bungee.event.EventPriority;
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Setting;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils;
import nl.chimpgamer.networkmanager.api.models.player.Player;

import java.util.Optional;

@RequiredArgsConstructor
public class JoinLeaveListener implements Listener {
    private final DiscordBot discordBot;

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onLogin(PostLoginEvent event) {
        final ProxiedPlayer proxiedPlayer = event.getPlayer();
        if (proxiedPlayer == null) {
            return;
        }
        Optional<Player> opPlayer = getDiscordBot().getNetworkManager().getPlayerSafe(proxiedPlayer.getUniqueId());
        if (!opPlayer.isPresent()) {
            getDiscordBot().getLogger().info("Could not load player...");
            return;
        }
        Player player = opPlayer.get();

        String discordId = this.getDiscordBot().getDiscordUserManager().getDiscordIdByUuid(player.getUuid());
        if (discordId == null) {
            return;
        }

        Preconditions.checkNotNull(this.getDiscordBot().getGuild(), "The discord bot has not been connected to a discord server. Connect it to a discord server.");

        Member member = this.getDiscordBot().getGuild().getMemberById(discordId);
        if (member == null) {
            return;
        }

        if (Setting.DISCORD_SYNC_USERNAME.getAsBoolean()) {
            Utils.setNickName(member, player.getName());
        }

        if (Setting.DISCORD_SYNC_RANKS_ENABLED.getAsBoolean()) {
            Utils.syncRanks(player);
        }
    }

    private DiscordBot getDiscordBot() {
        return discordBot;
    }
}

File: src/main/kotlin/nl/chimpgamer/networkmanager/extensions/discordbot/listeners/bungee/JoinLeaveListener.kt
package nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee

import net.md_5.bungee.api.event.PostLoginEvent
import net.md_5.bungee.api.plugin.Listener
import net.md_5.bungee.event.EventHandler
import net.md_5.bungee.event.EventPriority
import nl.chimpgamer.networkmanager.api.utils.Placeholders
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Setting
import nl.chimpgamer.networkmanager.extensions.discordbot.tasks.SyncRanksTask

class JoinLeaveListener(private val discordBot: DiscordBot) : Listener {

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onLogin(event: PostLoginEvent) {
        val proxiedPlayer = event.player ?: return
        val player = discordBot.networkManager.getPlayer(proxiedPlayer.uniqueId)
        if (player == null) {
            discordBot.logger.info("Could not load player...")
            return
        }
        val discordId = discordBot.discordUserManager.getDiscordIdByUuid(proxiedPlayer.uniqueId) ?: return
        checkNotNull(discordBot.guild) { "The discord bot has not been connected to a discord server. Connect it to a discord server." }
        val member = discordBot.guild.getMemberById(discordId) ?: return
        if (discordBot.settings.getBoolean(Setting.DISCORD_SYNC_USERNAME)) {
            val format = Placeholders.setPlaceholders(player, discordBot.settings.getString(Setting.DISCORD_SYNC_USERNAME_FORMAT))
            discordBot.discordManager.setNickName(member, format)
        }
        if (discordBot.settings.getBoolean(Setting.DISCORD_SYNC_RANKS_ENABLED)) {
            discordBot.scheduler.runSync(SyncRanksTask(discordBot, player))
        }
    }
}